// @backstage/plugin-search-backend-module-catalog/dist/alpha.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var backendPluginApi = require('@backstage/backend-plugin-api');
var alpha = require('@backstage/plugin-search-backend-node/alpha');
var pluginSearchBackendModuleCatalog = require('@backstage/plugin-search-backend-module-catalog');

const searchModuleCatalogCollator = backendPluginApi.createBackendModule(
  (options) => ({
    moduleId: "catalogCollator",
    pluginId: "search",
    register(env) {
      env.registerInit({
        deps: {
          config: backendPluginApi.coreServices.config,
          discovery: backendPluginApi.coreServices.discovery,
          tokenManager: backendPluginApi.coreServices.tokenManager,
          scheduler: backendPluginApi.coreServices.scheduler,
          indexRegistry: alpha.searchIndexRegistryExtensionPoint
        },
        async init({
          config,
          discovery,
          tokenManager,
          scheduler,
          indexRegistry
        }) {
          var _a;
          const defaultSchedule = {
            frequency: { minutes: 10 },
            timeout: { minutes: 15 },
            initialDelay: { seconds: 3 }
          };
          indexRegistry.addCollator({
            schedule: scheduler.createScheduledTaskRunner(
              (_a = options == null ? void 0 : options.schedule) != null ? _a : defaultSchedule
            ),
            factory: pluginSearchBackendModuleCatalog.DefaultCatalogCollatorFactory.fromConfig(config, {
              ...options,
              discovery,
              tokenManager
            })
          });
        }
      });
    }
  })
);

exports.searchModuleCatalogCollator = searchModuleCatalogCollator;
