// @keyv/redis/dist/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
const events_1 = __importDefault(require("events"));
const ioredis_1 = __importDefault(require("ioredis"));
class KeyvRedis extends events_1.default {
    constructor(uri, options) {
        super();
        this.ttlSupport = true;
        this.opts = {};
        this.opts.dialect = 'redis';
        if (typeof uri !== 'string' && uri.options && ('family' in uri.options || uri.isCluster)) {
            this.redis = uri;
        }
        else {
            options = { ...(typeof uri === 'string' ? { uri } : uri), ...options };
            // @ts-expect-error - uri is a string or RedisOptions
            this.redis = new ioredis_1.default(options.uri, options);
        }
        this.redis.on('error', (error) => this.emit('error', error));
    }
    _getNamespace() {
        return `namespace:${this.namespace}`;
    }
    async get(key) {
        const value = await this.redis.get(key);
        if (value === null) {
            return undefined;
        }
        return value;
    }
    async getMany(keys) {
        const rows = await this.redis.mget(keys);
        return rows;
    }
    async set(key, value, ttl) {
        if (value === undefined) {
            return undefined;
        }
        if (typeof ttl === 'number') {
            await this.redis.set(key, value, 'PX', ttl);
        }
        else {
            await this.redis.set(key, value);
        }
        await this.redis.sadd(this._getNamespace(), key);
    }
    async delete(key) {
        const items = await this.redis.del(key);
        await this.redis.srem(this._getNamespace(), key);
        return items > 0;
    }
    async deleteMany(key) {
        return this.delete(key);
    }
    async clear() {
        const keys = await this.redis.smembers(this._getNamespace());
        await this.redis.del([...keys, this._getNamespace()]);
    }
    async *iterator(namespace) {
        const scan = this.redis.scan.bind(this.redis);
        const get = this.redis.mget.bind(this.redis);
        let cursor = '0';
        do {
            // eslint-disable-next-line no-await-in-loop
            const [curs, keys] = await scan(cursor, 'MATCH', `${namespace}:*`);
            cursor = curs;
            if (keys.length > 0) {
                // eslint-disable-next-line no-await-in-loop
                const values = await get(keys);
                for (const [i] of keys.entries()) {
                    const key = keys[i];
                    const value = values[i];
                    yield [key, value];
                }
            }
        } while (cursor !== '0');
    }
    async has(key) {
        const value = await this.redis.exists(key);
        return value !== 0;
    }
    async disconnect() {
        return this.redis.disconnect();
    }
}
module.exports = KeyvRedis;
