// @backstage/plugin-search-backend/dist/alpha.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var backendPluginApi = require('@backstage/backend-plugin-api');
var backendCommon = require('@backstage/backend-common');
var pluginSearchBackendNode = require('@backstage/plugin-search-backend-node');
var alpha = require('@backstage/plugin-search-backend-node/alpha');
var router = require('./cjs/router-8576144f.cjs.js');
require('express-promise-router');
require('zod');
require('@backstage/errors');
require('@backstage/plugin-auth-node');
require('@backstage/plugin-permission-common');
require('lodash');
require('qs');
require('dataloader');

class SearchIndexRegistry {
  constructor() {
    this.collators = [];
    this.decorators = [];
  }
  addCollator(options) {
    this.collators.push(options);
  }
  addDecorator(options) {
    this.decorators.push(options);
  }
  getCollators() {
    return this.collators;
  }
  getDecorators() {
    return this.decorators;
  }
}
class SearchEngineRegistry {
  constructor() {
    this.searchEngine = null;
  }
  setSearchEngine(searchEngine) {
    if (this.searchEngine) {
      throw new Error("Multiple Search engines is not supported at this time");
    }
    this.searchEngine = searchEngine;
  }
  getSearchEngine() {
    return this.searchEngine;
  }
}
const searchPlugin = backendPluginApi.createBackendPlugin({
  pluginId: "search",
  register(env) {
    const searchIndexRegistry = new SearchIndexRegistry();
    env.registerExtensionPoint(
      alpha.searchIndexRegistryExtensionPoint,
      searchIndexRegistry
    );
    const searchEngineRegistry = new SearchEngineRegistry();
    env.registerExtensionPoint(
      alpha.searchEngineRegistryExtensionPoint,
      searchEngineRegistry
    );
    env.registerInit({
      deps: {
        logger: backendPluginApi.coreServices.logger,
        config: backendPluginApi.coreServices.config,
        permissions: backendPluginApi.coreServices.permissions,
        http: backendPluginApi.coreServices.httpRouter,
        searchIndexService: alpha.searchIndexServiceRef
      },
      async init({ config, logger, permissions, http, searchIndexService }) {
        let searchEngine = searchEngineRegistry.getSearchEngine();
        if (!searchEngine) {
          searchEngine = new pluginSearchBackendNode.LunrSearchEngine({
            logger: backendCommon.loggerToWinstonLogger(logger)
          });
        }
        const collators = searchIndexRegistry.getCollators();
        const decorators = searchIndexRegistry.getDecorators();
        await searchIndexService.start({
          searchEngine,
          collators,
          decorators
        });
        const router$1 = await router.createRouter({
          config,
          permissions,
          logger: backendCommon.loggerToWinstonLogger(logger),
          engine: searchEngine,
          types: searchIndexService.getDocumentTypes()
        });
        http.use(router$1);
      }
    });
  }
});

exports.searchPlugin = searchPlugin;
