// @backstage/plugin-search-backend-module-pg/dist/alpha.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var backendPluginApi = require('@backstage/backend-plugin-api');
var alpha = require('@backstage/plugin-search-backend-node/alpha');
var PgSearchEngine = require('./cjs/PgSearchEngine-b474d3ca.cjs.js');
require('@backstage/backend-common');
require('@backstage/plugin-search-backend-node');
require('uuid');

const searchModulePostgresEngine = backendPluginApi.createBackendModule({
  moduleId: "postgresEngine",
  pluginId: "search",
  register(env) {
    env.registerInit({
      deps: {
        searchEngineRegistry: alpha.searchEngineRegistryExtensionPoint,
        database: backendPluginApi.coreServices.database,
        config: backendPluginApi.coreServices.config
      },
      async init({ searchEngineRegistry, database, config }) {
        searchEngineRegistry.setSearchEngine(
          await PgSearchEngine.PgSearchEngine.fromConfig(config, {
            database
          })
        );
      }
    });
  }
});

exports.searchModulePostgresEngine = searchModulePostgresEngine;
